applications/DEMApplication/custom_utilities/tesselation_utilities.cpp

```diff
@@ -0,0 +1,805 @@
+    // Update voronoi diagram
+    if (mUpdateVoronoi)
+      UpdateVoronoi2D(r_modelpart, out, vorout);
+
+    // Update average porosity
+    if (mUpdatePorosiy)
+      UpdatePorosity2D(r_modelpart, out, vorout);
+
+    // Free memory
+    FreeTriangle(in);
+    FreeTriangle(out);
+    FreeTriangle(vorout);
+
+    ClearTriangle(in);
+    ClearTriangle(out);
+    ClearTriangle(vorout);
+
+    KRATOS_CATCH("")
+  }
+
+  //-----------------------------------------------------------------------------------------------------------------------
+  void TesselationUtilities::Tetrahedralization(ModelPart& r_modelpart)
+  {
+    KRATOS_TRY
+
+    int num_particles = r_modelpart.NumberOfElements();
+    if (num_particles < 4)
+      return;
+
+    // Build input
+    struct tetgenio in, out;
+    in.firstnumber    = 1;
+    in.mesh_dim       = 3;
+    in.numberofpoints = num_particles;
+    in.pointlist      = (double*)malloc(sizeof(double) * in.numberofpoints * 3);
+
+    #pragma omp parallel for schedule(dynamic, 100)
+    for (int i = 0; i < in.numberofpoints; i++) {
+      ModelPart::ElementsContainerType::iterator it = r_modelpart.GetCommunicator().LocalMesh().Elements().ptr_begin() + i;
+      ThermalSphericParticle<SphericParticle>& particle = dynamic_cast<ThermalSphericParticle<SphericParticle>&> (*it);
+
+      particle.mDelaunayPointListIndex = i;
+
+      const array_1d<double, 3>& coors = particle.GetParticleCoordinates();
+      in.pointlist[3 * i + 0] = coors[0];
+      in.pointlist[3 * i + 1] = coors[1];
+      in.pointlist[3 * i + 2] = coors[2];
+
+      if (mUpdateVoronoi)
+        particle.mNeighborVoronoiRadius.clear();
+    }
+
+    // Perform tetrahedralization
+    int fail = 0;
+    try {
+      tetrahedralize(&mSwitches[0], &in, &out);
+    }
+    catch (int error_code) {
+      fail = error_code;
+    }
+
+    if (fail || out.numberoftetrahedra == 0 || in.numberofpoints != out.numberofpoints) {
+      KRATOS_ERROR_IF(r_modelpart.GetProcessInfo()[TIME_STEPS] == 1) << "Fail to generate tetrahedralization!" << std::endl;
+      KRATOS_WARNING("DEM") << std::endl;
+      KRATOS_WARNING("DEM") << "Fail to generate tetrahedralization! Results from previous successful tetrahedralization will be used." << std::endl;
+      KRATOS_WARNING("DEM") << std::endl;
+      return;
+    }
+
+    // Update voronoi diagram
+    if (mUpdateVoronoi)
+      UpdateVoronoi3D(r_modelpart, out);
+
+    // Update average porosity
+    if (mUpdatePorosiy)
+      UpdatePorosity3D(r_modelpart, out);
+
+    KRATOS_CATCH("")
+  }
+
+  //-----------------------------------------------------------------------------------------------------------------------
+  /*
+  * Build a table for each particle with needed information from 2D voronoi diagram:
+  * column 1: neighbor particle IDs
+  * column 2: voronoi edge "radius" (length/2)
+  */
+  void TesselationUtilities::UpdateVoronoi2D(ModelPart& r_modelpart, struct triangulateio& out, struct triangulateio& vorout) {
```

Comment:
rOut, rVorOut

Reply:
Name changed in every function of the class